src/decoders/belief_propogation.jl

```diff
@@ -0,0 +1,139 @@
+using SparseArrays
+
+include("abstract_decoder.jl")
+
+struct BeliefPropagationSetup
+  "Log probabilities"
+  log_probabs::Vector{Float64}
+  
+  "Channel probabilities"
+  channel_probs::Vector{Float64}
+  
+  "Bit to check message matrix"
+  bit_2_check::Matrix{Float64}
+
+  "Check to bit message matrix"
+  check_2_bit::Matrix{Float64}
+
+  "predicted error for each bit"
+  err::Vector{Float64}
+end 
+
+
+struct BeliefPropagationDecoder{T} <: AbstractDecoder
+  "Parity check matrix"
+  H::T
+
+  "Physical error rate"
+  per::Float64
+
+  "Number of max iterations of Belief propagation decoder"
+  max_iters::Int
+end
+
+struct BeliefPropagationDecoderSparse{T} <: AbstractDecoder
+  "Parity check matrix"
+  H::T
+
+  "Physical error rate"
+  per::Float64
+
+  "Number of max iterations of Belief propagation decoder"
+  max_iters::Int
+  
+  "Num of parity checks i.e. number of rows of parity check matrix"
+  s::Int
+  
+  "Num of bits in the code i.e number of columns of parity check matrix"
+  n::Int
+  
+  sparse_H::SparseArrays.SparseMatrixCSC{Bool,Int}
+  
+  sparse_HT::SparseArrays.SparseMatrixCSC{Bool,Int}
```

Comment:
Could you add docstrigs for these too?

When should one use `BeliefPropagationDecoder` and when should they use `BeliefPropagationDecoderSparse`? Could we make it into one single type, probably parameterized?

Reply:
If I am reading this correctly, you use `BeliefPropagationDecoderSparse` only internally, constructing it out of the public `BeliefPropagationDecoder`.

I think it would be better to have only `BeliefPropagationDecoder` with a custom constructor that populates the extra fields.

i.e.
```
struct BeliefPropagationDecoder
    H...
    per...
    max...
    sparse...
end

function BeliefPropagationDecoder(H, per, max_iters)
    sparse_H = ...
    sparse_Ht = ...
    return BeliefPropagationDecoder(H, per, max_iters, sparse_H, sparse_Ht)
end
```

Reply:
Actually, you should probably completely drop `H` from the struct, no point to save it, especially if it is very large. It can just be an argument for the constructor.